src/dft.cpp

```diff
@@ -1375,4 +1388,94 @@ void fields::get_mode_mode_overlap(void *mode1_data, void *mode2_data, dft_flux
   get_overlap(mode1_data, mode2_data, flux, 0, overlaps);
 }
 
+// return the size of the dft monitor
+std::vector<int> fields::dft_monitor_size(dft_fields fdft, const volume &where, component c){
+  ivec min_corner, max_corner;
+  int rank, reduced_rank;
+  direction dirs[3], reduced_dirs[3];
+  size_t array_size, bufsz, reduced_grid_size, reduced_stride[3], stride[3];
+  dft_chunk *chunklists[1];
+  chunklists[0] = fdft.chunks;
+
+  size_t *dims = get_dims(chunklists, 1, c, min_corner, max_corner, array_size, bufsz, rank, dirs);
+  size_t *reduced_dims = reduce_array_dimensions(where, rank, reduced_rank, reduced_grid_size, dims, stride, reduced_stride, dirs, reduced_dirs);
+
+  std::vector<int> monitor_array = {1, 1, 1}; // lengths of the dft monitor along the three dimensions
+  for (int nd = 0; nd < 3; ++nd) if (nd < reduced_rank) monitor_array[nd] = reduced_dims[nd];
+
+  return monitor_array;
+}
+
+std::vector<struct sourcedata> dft_fields::fourier_sourcedata(const volume &where, component c, fields &f, const std::complex<double>* dJ){
+  const size_t Nfreq = freq.size();
+
+  ivec min_corner, max_corner;
+  int rank, reduced_rank;
+  direction dirs[3], reduced_dirs[3];
+  size_t array_size, bufsz, reduced_grid_size, reduced_stride[3], stride[3];
+  dft_chunk *chunklists[1];
+  chunklists[0] = chunks;
+  
+  size_t *dims = f.get_dims(chunklists, 1, c, min_corner, max_corner, array_size, bufsz, rank, dirs);
+  size_t *reduced_dims = reduce_array_dimensions(where, rank, reduced_rank, reduced_grid_size, dims, stride, reduced_stride, dirs, reduced_dirs);
+  size_t monitor_array[3] = {1, 1, 1}; // lengths of the dft monitor along the three dimensions
```

Comment:
With the above changes, you should be able to get rid of `monitor_array` and just use `reduced_grid_size` directly.